### ISSUE TITLE
The change to deserialising variables by ID instead of by value broke backwards compatibility of serialised Blockly XML

### ISSUE THREAD

**Issue by timdawborn** [google/blockly#1744]
#1531 introduced referring to variables by ID instead of by value (yay!). It appears to have broken the deserialisation of old Blockly XML however. 

In [this mailing list thread](https://groups.google.com/forum/#!topic/blockly/VWwmJfZlvM0), @NeilFraser and @rachel-fenichel  claim that the Blockly XML representation is backwards compatible in its representation:

> The XML format has slowly grown over time, however always in a reverse-compatible manner.  XML from the very first Blockly release more than five years ago will still load flawlessly today. 

> As Neil said, our XML changes are all backwards compatible.

It's true that the representation is backwards compatible, but this changeset appears to have caused a non-backwards compatible change in the semantics of what's serialised.

We have *heaps* of serialised Blockly XML from all stages of the Blockly codebase. I'm now trying to work out whether we need to go and correct all of it for these new semantics or whether this is an unexpected change by the core Blockly team and they plan to fix the backwards compatible issue.

Here's an example of what I'm referring to. This Blockly XML was generated mid 2017 on our end, so would have been using a Blockly version from somewhere around the start of 2017 (circa 24ebbcaad09d079468e17b600f8a742b676f5c54). I could find the exact revision if necessary.

```xml
<blockly>
  <variables>
    <variable type="" id="1">line_length</variable>
  </variables>
  <block type="variables_set" id="2" inline="true" x="10" y="10">
    <field name="VAR" id="3" variabletype="">line_length</field>
    <value name="VALUE">
      <block type="math_number" id="4">
        <field name="NUM">100</field>
      </block>
    </value>
    <next>
      <block type="turtle_forward" id="5">
        <value name="DISTANCE">
          <block type="variables_get" id="6">
            <field name="VAR" id="7" variabletype="">line_length</field>
          </block>
        </value>
        <next>
          <block type="turtle_turn_simple" id="8">
            <field name="DIRECTION">LEFT</field>
            <next>
              <block type="turtle_forward" id="9">
                <value name="DISTANCE">
                  <block type="variables_get" id="10">
                    <field name="VAR" id="11" variabletype="">line_length</field>
                  </block>
                </value>
              </block>
            </next>
          </block>
        </next>
      </block>
    </next>
  </block>
</blocky>
```

Note that the `line_length` variable is defined in the `<variables>` tree with ID 1 and then again in the `<variables_set>` tree with ID 3. From what I can tell looking at the changes over time, this was because, once upon a time, the XML serialisation always used the Block's ID as the `id` attribute on a field: see https://github.com/google/blockly/blob/24ebbcaad09d079468e17b600f8a742b676f5c54/core/xml.js#L82

#1531 changed the semantics of this `id` attribute on serialised `<field>` nodes, expecting it to be the variable ID instead of the block ID. For comparison, the same Blockly program as above serialised using Blockly as it is today yields:

```xml
<xml xmlns="http://www.w3.org/1999/xhtml">
  <variables>
    <variable type="" id="iy}4v*dDKKU2|zYDWyjq">line_length</variable>
  </variables>
  <block type="variables_set" id="J3wMH6btk8+OIX^#Kq)Z" x="20" y="12">
    <field name="VAR" id="iy}4v*dDKKU2|zYDWyjq" variabletype="">line_length</field>
    <value name="VALUE">
      <block type="math_number" id="Cp:B~m,UR`#,2`b){b6q">
        <field name="NUM">100</field>
      </block>
    </value>
    <next>
      <block type="turtle_forward" id="0;~@6}gpILX6v(cCa9XQ">
        <value name="DISTANCE">
          <block type="variables_get" id="|Zn-mli:f2hO[k1Tak;2">
            <field name="VAR" id="iy}4v*dDKKU2|zYDWyjq" variabletype="">line_length</field>
          </block>
        </value>
        <next>
          <block type="turtle_turn_simple" id="Qx=aBrvcmvj`5Ho6wX3k">
            <field name="DIRECTION">LEFT</field>
            <next>
              <block type="turtle_forward" id="@BnWj[CfT#)7JE9`;B;%">
                <value name="DISTANCE">
                  <block type="variables_get" id="PA+!FqRis%gXZl/iF#5/">
                    <field name="VAR" id="iy}4v*dDKKU2|zYDWyjq" variabletype="">line_length</field>
                  </block>
                </value>
              </block>
            </next>
          </block>
        </next>
      </block>
    </next>
  </block>
</xml>
```

Note that the `id` attribute is the same for all uses of the `line_length` variable.

The semantic change of the `id` attribute would be okay except that it throws an exception during deserialisation due to the variable IDs no longer matching: https://github.com/google/blockly/pull/1531/files#diff-5d7a3b9d67dc9e0425dd115a1aac67d1R174 . As such, attempting to load the old serialised XML dies at runtime:
<img width="1251" alt="screen shot 2018-03-29 at 10 15 10" src="https://user-images.githubusercontent.com/812019/38061228-3c9c5f66-333a-11e8-836d-becaa7ea9515.png">

So, back to my particular issue. We have *heaps* of serialised Blockly XML from all stages of the Blockly codebase. I'm now trying to work out whether we need to go and adjust all of it to account for this semantic change to the `id` attribute of variable `<field>`, or whether this is an unexpected change by the core Blockly team and they plan to fix the backwards compatible issue.

**Comment by rachel-fenichel** [comment]
[Oh dear](https://i.imgur.com/1dj95zH.gif).

That sounds unexpected.  I put a ton of work into trying to keep it backwards compatible.  I'll take a look.

**Comment by timdawborn** [comment]
There's a relatively simple workaround which is to remove the exception being thrown when the IDs conflict, but I assume it is there for a reason. It was added in [d903b5e86ba247dbecd31be16b4513eaa94ee889](https://github.com/marisaleung/blockly/commit/d903b5e86ba247dbecd31be16b4513eaa94ee889#diff-84b5fa00c539dcb35c0fc1e659c2a9d1R348) whose commit message looks automated.

**Comment by rachel-fenichel** [comment]
Hm, it can't be from the line that you pointed to, because that's setting the id on the <block> element, not the <field> element.  Here's the rest of the code around it:
```
Blockly.Xml.blockToDom = function(block, opt_noId) {
  var element = goog.dom.createDom(block.isShadow() ? 'shadow' : 'block');
  element.setAttribute('type', block.type);
  if (!opt_noId) {
    element.setAttribute('id', block.id);
  }
  ...
```

And here's the `fieldToDom` function from that revision: 
```
  function fieldToDom(field) {
    if (field.name && field.EDITABLE) {
      var container = goog.dom.createDom('field', null, field.getValue());
      container.setAttribute('name', field.name);
      element.appendChild(container);
    }
  }
```

In your XML it looks like variable fields are getting ids, but other fields aren't (e.g. `<field name="NUM">100</field>`).  That makes me think that it's from an early attempt to get variables by ID working.

Helpful info:
- Exact revision hash.  I'm going to do some spelunking.
- Whether that code was on develop or master when you pulled.

Cheers,
Rachel

**Comment by timdawborn** [comment]
Still trying to find the exact revision. We've changed our approach to our fork over time as well, so working out what the upstream revision would have been is taking time.

It would have been off `develop` though, not master. If you reckon that this fact has caused us to shoot ourselves in the foot, that's fine. I'm happy to eat that and we can work out how to correct our serialised XML. That's what we get to living on the bleeding edge.

**Comment by rachel-fenichel** [comment]
Okay, if you branched from develop our guarantee about backwards compatibility is not nearly as strong.  It looks like you caught one of the intermediate states on this code, when we had merged a feature but not worked out all of the bugs.

No need to work out the upstream revision, then.  

Let's sort out how to correct that XML, and it should hopefully be just a blip.

**Comment by rachel-fenichel** [comment]
If it's feasible for you, the way to correct this would be to load and then save any suspect XML, using a tweaked loader that ignores these problems.  That should "clean" all of your XML.  Finding impacted projects is probably time-based.  Once every project from the broken period is fixed, the normal load and save paths should be fine.

Assuming that that's possible, and that you aren't doing anything overly fancy with variables (not having multiple variables by the same name with different types, mostly), the change is simple.

In `Blockly.Xml.domToFieldVariable_` change this line:
```
var variable = Blockly.Variables.getOrCreateVariablePackage(workspace, xml.id,
      text, type);
```

to this line:

```
var variable = Blockly.Variables.getOrCreateVariablePackage(workspace, null,
      text, type);
```

**Comment by rachel-fenichel** [comment]
Verified with this XML on the playground:

```
<xml xmlns="http://www.w3.org/1999/xhtml">
  <variables>
    <variable type="" id="1">bar</variable>
    <variable type="" id="2">baz</variable>
  </variables>
  <block type="variables_set" id="a" x="13" y="13">
    <field name="VAR" id="3" variabletype="">bar</field>
  </block>
  <block type="variables_set" id="b" x="13" y="63">
    <field name="VAR" id="1" variabletype="">bar</field>
  </block>
  <block type="variables_set" id="c" x="13" y="113">
    <field name="VAR" id="1" variabletype="">bar</field>
  </block>
  <block type="variables_get" id="d" x="13" y="163">
    <field name="VAR" id="2" variabletype="">baz</field>
  </block>
</xml>
```
The error is in the first reference to bar, where it's using ID 3 instead of 1.  This fails on develop but works in my tweaked code, and then serializes correctly.

**Comment by timdawborn** [comment]
Yup, thanks. I had come to the conclusion of the same patch too.

I got massively distracted trying to repro the old serialised XML I quoted originally, but I'm failing to do so. I can't find any revision where Block IDs still used `Blockly.Block.genUid` (counter based) instead of `Blocky.utils.genUid` (UUID-style), *and* had the `<variables>` element in the DOM. The `<variables>` element seems to have been added in ce4a84b3cf72163e97a165957de2bf78a5dd11b0 which is way after the UUID-style block IDs came into place. I have no idea what's happened here.

I guess I'll try and work out what percentage of our Blockly fails to load, and work out whether to patch our Blockly fork, or just adjust the serialised Blockly.

Thanks for looking into this anyway.
